### ISSUE TITLE
TypeError: $rdf.fetcher.lookUpThing is not a function

### ISSUE THREAD

**Issue by ckristo** [linkeddata/rdflib.js#111]
Hi,

I updated my rdflib.js example from version 0.2.6 to 0.5 and encounter the following error:

````
$rdf.fetcher.lookUpThing(st.subject, 'sparql:' + st.subject)
                     ^

TypeError: $rdf.fetcher.lookUpThing is not a function
    at Object.$rdf.SPARQLToQuery (/Users/christoph/Documents/Ausbildung/TU/033532 Medieninformatik/Bakk-Arbeit/Arbeit/examples/js/rdflib.js-example/node_modules/rdflib/dist/rdflib-node.js:6751:22)
````

It seems that `$rdf.fetcher` is a function, and not an object, see https://ckristo.net/owncloud/index.php/s/mcV0FZCP838dfJz

As workaround, setting `$rdf.fetcher` to null causes the faulty function call to be skipped.

BR, Chris

**Comment by ckristo** [comment]
cf. https://github.com/linkeddata/rdflib.js/issues/70

**Comment by ckristo** [comment]
Shouldn't `testMode` if set to `false` disable resource fetching? Why does `$rdf.SPARQLToQuery` then even try?

**Comment by timbl** [comment]
Yes, $rdf.fetcher is a function which returns you a new fetcher but you must  pass it a graph.   The two are linked together.   
```
   var g = $rdf.graph()
   var f = $rdf.fetcher(g)

g.add() ...
f.load() ...

```
You can think of $rdf.fetcher(g)  as a shorthand for new $rdf.Fetcher(g)   -- does this make sense?

DO NOT do:
```
 $rdf.fetcher = null; /*   BAD */ 
```
that breaks the system!

**Comment by ckristo** [comment]
Hi,

I do not use `$rdf.fetcher` at all, the error is produced by rdflib.js's `SPARQLToQuery` method (see the example included above).

There is a check which tests if `$rdf.fetcher` is set in the code, and if you set it to null, the erroneous function call `.lookUpThing()` gets skipped. Sure, it is not a good idea to set it to null, but in my case this is the only way to get my example to work.

This is a bug and should stay open imo.

BR, Chris

**Comment by ckristo** [comment]
`$rdf.fetcher` is used as some kind of factory method, right? Why does `SPARQLToQuery` use it as object? I suppose it expects `$rdf.fetcher` be a fetcher instance.

There are two possibilities I see: either create a new one with the factory method, or use an existing one. The `SPARQLToQuery` method takes `kb`, an IndexedFormula, as argument which might hold a fetcher instance as `kb.fetcher` property.

Due to the fact that the code of `SPARQLToQuery` is quite complex, opinions on that matter from the original author(s) would be nice for preparing a PR.

BR, Chris

**Comment by jeswr** [comment]
Verified against rdflib v2.4.0 as part of the maintainer-signed-off triage (plan #824): the original TypeError no longer reproduces — SPARQLToQuery now uses the store's own fetcher, guarded by `if (kb.fetcher)` (src/sparql-to-query.js), so calling it without a fetcher attached is safe and testMode returns early as expected. Closing as fixed by the fetcher/query rewrite; the companion PR #113 is likewise obsolete. Please reopen with a fresh repro if you still hit this on a current release.
